src/tree/updater_quantile_hist.cc

```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
   }
 }
 
-void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
-                                        const GHistIndexBlockMatrix& gmatb,
-                                        const ColumnMatrix& column_matrix,
-                                        HostDeviceVector<GradientPair>* gpair,
-                                        DMatrix* p_fmat,
-                                        RegTree* p_tree) {
-  double gstart = dmlc::GetTime();
+void QuantileHistMaker::Builder::SyncHistograms(
+    int &starting_index,
+    int &sync_count,
+    RegTree *p_tree) {
+  this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
+  // use substracttricks
+  for (auto local_it = nodes_to_derive_.begin(); local_it != nodes_to_derive_.end(); local_it++) {
+    hist_.AddHistRow(local_it->first);
+    SubtractionTrick(hist_[local_it->first], hist_[local_it->second],
+                     hist_[(*p_tree)[local_it->first].Parent()]);
+  }
+}
 
-  int num_leaves = 0;
-  unsigned timestamp = 0;
+void QuantileHistMaker::Builder::BuildLocalHistograms(
+    int &starting_index,
+    int &sync_count,
+    const GHistIndexMatrix &gmat,
+    const GHistIndexBlockMatrix &gmatb,
+    RegTree *p_tree,
+    const std::vector<GradientPair> &gpair_h) {
+  // 0. build histogram for the nodes in current level
+  for (size_t k = 0; k < qexpand_depth_wise_.size(); k++) {
+    int nid = qexpand_depth_wise_[k].nid;
+    RegTree::Node &node = (*p_tree)[nid];
+    if (rabit::IsDistributed()) {
+      if (node.IsRoot() || node.IsLeftChild()) {
+        // in distributed setting, we always calcuate from left child or root node
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        if (!node.IsRoot() && node.IsLeftChild()) {
+          nodes_to_derive_[(*p_tree)[node.Parent()].RightChild()] = nid;
+        }
+        sync_count++;
+        starting_index = std::min(starting_index, nid);
+      }
+    } else {
+      if (!node.IsRoot() && node.IsLeftChild() &&
+          (row_set_collection_[nid].Size() <
+           row_set_collection_[(*p_tree)[node.Parent()].RightChild()].Size())) {
+        hist_.AddHistRow(nid);
```

Comment:
actually no, for example 

if this is a left child and its number of samples is larger than its sibling's, we should do nothing and defer it to the time when we visit the sibling

if we move these lines out of test branches, it will be added in anyway